corporate/tests/test_stripe.py

```diff
@@ -2887,6 +2887,172 @@ def test_downgrade_free_trial(self) -> None:
+            with self.assertLogs("corporate.stripe", "INFO") as m:
+                with patch("corporate.lib.stripe.timezone_now", return_value=self.now):
+                    response = self.client_patch(
+                        "/json/billing/plan",
+                        {"status": CustomerPlan.DOWNGRADE_AT_END_OF_FREE_TRIAL},
+                    )
+                    stripe_customer_id = Customer.objects.get(realm=user.realm).id
+                    new_plan = get_current_plan_by_realm(user.realm)
+                    assert new_plan is not None
+                    expected_log = f"INFO:corporate.stripe:Change plan status: Customer.id: {stripe_customer_id}, CustomerPlan.id: {new_plan.id}, status: {CustomerPlan.DOWNGRADE_AT_END_OF_FREE_TRIAL}"
+                    self.assertEqual(m.output[0], expected_log)
+                    self.assert_json_success(response)
+            plan.refresh_from_db()
+            self.assertEqual(plan.next_invoice_date, free_trial_end_date)
+            self.assertEqual(get_realm("zulip").plan_type, Realm.PLAN_TYPE_STANDARD)
+            self.assertEqual(plan.status, CustomerPlan.DOWNGRADE_AT_END_OF_FREE_TRIAL)
+            self.assertEqual(plan.licenses(), self.seat_count)
+            self.assertEqual(plan.licenses_at_next_renewal(), None)
+
+            with patch("corporate.lib.stripe.timezone_now", return_value=self.now):
+                mock_customer = Mock(email=user.delivery_email)
+                mock_customer.invoice_settings.default_payment_method = Mock(
+                    spec=stripe.PaymentMethod, type=Mock()
+                )
+                with patch("corporate.lib.stripe.stripe_get_customer", return_value=mock_customer):
+                    response = self.client_get("/billing/")
+                    self.assert_in_success_response(
+                        [
+                            "Your organization will be downgraded to <strong>Zulip Cloud Free</strong> at the end of the free trial",
+                            "<strong>March 2, 2012</strong>",
+                        ],
+                        response,
+                    )
+
+            # Verify that we still write LicenseLedger rows during the remaining
+            # part of the cycle
+            with patch("corporate.lib.stripe.get_latest_seat_count", return_value=20):
+                update_license_ledger_if_needed(user.realm, self.now)
+            self.assertEqual(
+                LicenseLedger.objects.order_by("-id")
+                .values_list("licenses", "licenses_at_next_renewal")
+                .first(),
+                (20, 20),
+            )
+
+            # Verify that we don't invoice them for the additional users during free trial.
+            from stripe import Invoice
+
+            Invoice.create = lambda **args: None  # type: ignore[assignment] # cleaner than mocking
+            Invoice.finalize_invoice = lambda *args: None  # type: ignore[assignment] # cleaner than mocking
```

Comment:
I'm skeptical that this non-mocking approach will not corrupt the `stripe` library in a way that breaks future tests.

Reply:
Sounds good.